### PR TITLE
feat(tree-view): add tree-view pattern

### PR DESCRIPTION
**Pull request by angelcabo**
Adds the sp-tree-view component pattern.

## Description

Implements a tree-view pattern and:
* Allows the use of shiftKey & metaKey while clicking tree-view-items to allow multi-selecting items and contiguous selection within a root-level tree-view component.
* Supports keyboard navigation and the recommended selection model as documented here: https://www.w3.org/TR/wai-aria-practices/#keyboard-interaction-22
* Implements all Tree View variants as documented here: https://opensource.adobe.com/spectrum-css/treeview.html

## Related Issue

https://github.com/adobe/spectrum-web-components/issues/1222

## Motivation and Context

This change implements the tree-view pattern to support multi-select and keyboard navigation as described here: https://www.w3.org/TR/wai-aria-practices/#TreeView, as well as all of the patterns on the spectrum-css site: https://opensource.adobe.com/spectrum-css/treeview.html

## How Has This Been Tested?

Yes. Tested manually with storybook and automated tests.

## Screenshots (if appropriate):

Screen Shot 2021-05-24 at 10 48 50 AM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by angelcabo** [comment]
@Westbrook I cleaned up the implementation from here: https://github.com/adobe/spectrum-web-components/pull/1423, addressed your feedback re: accessibility (let me know if I missed anything) and tried covering all tree-view patterns documented on the spectrum-css site. Also, i went through the https://www.w3.org/TR/wai-aria-practices/#TreeView guide and implemented the recommended keyboard navigation/selection model documented there. Let me know how things are looking. I can also take a stab at updating the relevant docs sections if you think this is in the right direction.

File: package.json [hunk 1]
```diff
@@ -110,6 +110,7 @@
         "@types/command-line-usage": "^5.0.1",
         "@types/common-tags": "^1.4.0",
         "@types/lunr": "^2.3.2",
+        "@types/mocha": "^8.2.2",
```
Comment (angelcabo) [inline package.json:113]: Without this, my editor was having trouble finding the `describe` and `it` globals provided by mocha.

**Comment by Westbrook** [comment]
@angelcabo I point this PR to the `westbrook/treeview` branch that I had been working on today. I've just rebased it to `main`, so it would be great if you could rebase to it so we can clarify any conflicts to there. If I'm not able to set aside some time for this tomorrow, I'll try to get into it first thing next week.

**Comment by angelcabo** [comment]
Thanks for taking a look, @Westbrook. I resolved the conflicts.

I had originally only intended on submitting a few changes to move the needle on the multi-select tree-view functionality after I saw you moved the tree-view issue out of "in progress" status, which was why I had opened it against your WIP branch in the first place. I hope I didn't step on your toes with this PR! From your initial feedback though, I started getting into tackling all of the accessibility issues, and pulling that thread led to me tackling most of the rest of the documented tree-view patterns, which was why I opened a new branch against main after your initial round of feedback.

I know this PR was a little weird given it was based on changes you already started, but in the future, are new PRs typically pulled into next or main first?

**Comment by Westbrook** [comment]
PRs generally target `main`. In this case, I'm not completely sure that I'm done with the work on my end, so pointing to to a staging location will allow me to sew up anything I've overlooked, and get a review on those changes, etc. Sometimes we spin up `next` as a staging ground for a long running project that requires multiple PRs, but try not to get into that level of bifurcation on a regular basis.

**Comment by angelcabo** [comment]
@Westbrook FYI, I've been trying to keep this PR up-to-date and tracking with your tree-view feat branch to avoid any merge conflicts. Let me know if there's anything I can do to help move things along.

**Review by Westbrook — APPROVED** [review]
@angelcabo thanks so much for this! This is looking so good.

There are some tiny nuances here and there that I might massage as I get my branch ready to go into `main` (think like event options, comment locations, etc.) but this will be a huge help in getting it over the wire.

My apologies for this taking so long to get to, but LGTM!

**Comment by angelcabo** [comment]
Awesome! Glad to hear @Westbrook. Thank you for providing the initial feedback and guidance. And no worries about timing, I know you guys are very busy with the rest of the library. If there's anything I can do to help with the feature or your branch once you get time to touch it up, please don't hesitate to reach out, either here or in the Slack project channel/DMs.

**Pull request merged**